src/bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,363 @@
+from abc import abstractmethod
+from typing import Dict, List
+
+import numpy as np
+import pandas as pd
+from event_model import DocumentRouter, RunRouter
+from event_model.documents import EventDescriptor, StreamDatum, StreamResource
+from pydantic.utils import deep_update
+from tiled.client import from_profile, from_uri
+from tiled.client.utils import handle_error
+from tiled.structures.array import ArrayStructure, BuiltinDtype
+from tiled.structures.core import Spec, StructureFamily
+from tiled.structures.data_source import Asset, DataSource, Management
+from tiled.structures.table import TableStructure
+from tiled.utils import safe_json_dump
+
+MIMETYPE_LOOKUP = {
+    "hdf5": "application/x-hdf5",
+    "ADHDF5_SWMR_STREAM": "application/x-hdf5",
+    "AD_HDF5_SWMR_SLICE": "application/x-hdf5",
+    "AD_TIFF": "multipart/related;type=image/tiff",
+}
+DTYPE_LOOKUP = {"number": "<f8", "array": "<f8", "boolean": "bool", "string": "str", "integer": "int"}
+
+
+# TODO (probably in a future PR?):
+# Move these into external repo (probably area-detector-handlers)
+# and use the existing handler discovery mechanism.
+
+
+class StreamHandlerBase:
+    """
+    A Handler consumes documents from Bluesky. It composes details (DataSource and its Assets) that will go into
+    the Tiled database. Each Handler (more specifically, a StreamHandler) is instantiated per a Stream Resource.
+
+    Tiled Adapters will later use this to read the data, with good random access and bulk access support.
+
+    We put this code into handlers so that additional, possibly very unusual, formats can be supported by users
+    without getting a PR merged into Bluesky or Tiled.
+
+    The STREAM_HANDLER_REGISTRY (see example below) and the Tiled catalog paramter adapters_by_mimetype can be used
+    together to support:
+        - ingesting a new mimetype from Bluesky documents and generating DataSource and Asset with appropriate
+          parameters (the handler's job);
+        - Interpreting those DataSource and Asset parameters to do I/O (the adapter's job).
+    """
+
+    mimetype: str = ""
+
+    def __init__(self, sres: StreamResource, desc: EventDescriptor):
+        assert (
```

Comment:
I have read that `assert` should not be used to validate user inputs. Instead a `ValueError` should be raised.

One good reason for this is that python can be run in a mode where all `assert` statements are ignored, for performance.

Reply:
Thank you for noting this, Dan -- this assertion was something else I wanted to ask you.
I read similar notes too somewhere, but I've also seen `assert` used to raise exceptions throughout the Bluesky code quite often; I thought it was a matter of style, but I might be missing some subtle differences here. Personally, I'd also prefer raising an exception explicitly, and I have refactored this part.

Reply:
Assert to check conditions that could only be false due a fully internal bug is OK. I'm mostly concerned with asserts that could be false due to erroneous user inputs.

Reply:
Fixed.